extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSource.java

```diff
@@ -0,0 +1,204 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.delta.input;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Iterators;
+import com.google.common.primitives.Ints;
+import io.delta.kernel.Scan;
+import io.delta.kernel.Snapshot;
+import io.delta.kernel.Table;
+import io.delta.kernel.TableNotFoundException;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.FilteredColumnarBatch;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.defaults.client.DefaultTableClient;
+import io.delta.kernel.internal.util.Utils;
+import io.delta.kernel.utils.CloseableIterator;
+import org.apache.druid.data.input.InputFormat;
+import org.apache.druid.data.input.InputRowSchema;
+import org.apache.druid.data.input.InputSource;
+import org.apache.druid.data.input.InputSourceReader;
+import org.apache.druid.data.input.InputSplit;
+import org.apache.druid.data.input.SplitHintSpec;
+import org.apache.druid.data.input.impl.SplittableInputSource;
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.utils.Streams;
+import org.apache.hadoop.conf.Configuration;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Input source to ingest data from a Delta Lake.
+ * This input source reads the latest snapshot from a Delta table specified by {@code tablePath} parameter.
+ * Note: the kernel table API only supports reading from the latest snapshot.
+ */
+public class DeltaInputSource implements SplittableInputSource<DeltaSplit>
+{
+  public static final String TYPE_KEY = "delta";
+
+  @JsonProperty
+  private final String tablePath;
+
+  @JsonProperty
+  @Nullable
+  private final DeltaSplit deltaSplit;
+
+  @JsonCreator
+  public DeltaInputSource(
+      @JsonProperty("tablePath") String tablePath,
+      @JsonProperty("deltaSplit") @Nullable DeltaSplit deltaSplit
+  )
+  {
+    if (tablePath == null) {
+      throw InvalidInput.exception("tablePath cannot be null");
+    }
+    this.tablePath = tablePath;
+    this.deltaSplit = deltaSplit;
+  }
+
+  @Override
+  public boolean needsFormat()
+  {
+    // Only support Parquet
+    return false;
+  }
+
+  @Override
+  public InputSourceReader reader(
+      InputRowSchema inputRowSchema,
+      @Nullable InputFormat inputFormat,
+      File temporaryDirectory
+  )
+  {
+    Configuration hadoopConf = new Configuration();
+    TableClient tableClient = DefaultTableClient.create(hadoopConf);
+    try {
+      final Row scanState;
+      final List<Row> scanRowList;
+
```

Comment:
can you add some comments describing the implementation?

Reply:
Added javadocs to this function